### PR TITLE
Alfoa/add loader pickled rom

### LINKED ISSUE #591: Make the importing of pickled ROMs in external Python easier
--------
Issue Description
--------
##### What did you expect to see happen?
Make the importing of pickled ROMs in external Python easier. 

##### Do you have a suggested fix for the development team?

Develop a module for importing pickled ROMs in external codes, exposing, correctly , the API

##### Please attach the input file(s) that generate this error.  The simpler the input, the faster we can find the issue.

----------------
For Change Control Board: Issue Review
----------------
This review should occur before any development is performed as a response to this issue.
- [x] 1. Is it tagged with a type: defect or improvement?
- [x] 2. Is it tagged with a priority: critical, normal or minor?
- [x] 3. If it will impact requirements or requirements tests, is it tagged with requirements?
- [x] 4. If it is a defect, can it cause wrong results for users? If so an email needs to be sent to the users.
- [x] 5. Is a rationale provided? (Such as explaining why the improvement is needed or why current code is wrong.)

-------
For Change Control Board: Issue Closure
-------
This review should occur when the issue is imminently going to be closed.
- [x] 1. If the issue is a defect, is the defect fixed?
- [x] 2. If the issue is a defect, is the defect tested for in the regression test system?  (If not explain why not.)
- [x] 3. If the issue can impact users, has an email to the users group been written (the email should specify if the defect impacts stable or master)?
- [x] 4. If the issue is a defect, does it impact the latest stable branch? If yes, is there any issue tagged with stable (create if needed)?
- [x] 5. If the issue is being closed without a merge request, has an explanation of why it is being closed been provided?

### PR DESCRIPTION
**Pull request by alfoa**
--------
Pull Request Description
--------
##### What issue does this change request address? (Use "#" before the issue to link it, i.e., #42.)
Closes #591

##### What are the significant changes in functionality due to this change request?
Just added a script for exporting pickled ROMs

----------------
For Change Control Board: Change Request Review
----------------
The following review must be completed by an authorized member of the Change Control Board.
- [x] 1. Review all computer code.
- [x] 2. If any changes occur to the input syntax, there must be an accompanying change to the user manual and xsd schema. If the input syntax change deprecates existing input files, a conversion script needs to be added (see Conversion Scripts).
- [x] 3. Make sure the Python code and commenting standards are respected (camelBack, etc.) - See on the [wiki](https://github.com/idaholab/raven/wiki/RAVEN-Code-Standards#python) for details.
- [x] 4. Automated Tests should pass, including run_tests, pylint, manual building and xsd tests. If there are changes to Simulation.py or JobHandler.py the qsub tests must pass.
- [x] 5. If significant functionality is added, there must  be tests added to check this. Tests should cover all possible options.  Multiple short tests are preferred over one large test. If new development on the internal JobHandler parallel system is performed, a cluster test must be added setting, in <RunInfo> XML block, the node ```<internalParallel>``` to True.
- [x] 6. If the change modifies or adds a requirement or a requirement based test case, the Change Control Board's Chair or designee also needs to approve the change.  The requirements and the requirements test shall be in sync.
- [x] 7. The merge request must reference an issue.  If the issue is closed, the issue close checklist shall be done.
- [x] 8. If an analytic test is changed/added is the the analytic documentation updated/added?

### CODE REVIEW

File: tests/framework/ROM/pickleTests/tests [hunk 1]
```diff
@@ -22,4 +22,10 @@
     input = 'not_a_rom_error.xml'
     expect_err = 'RuntimeError: Pickled object in'
   [../]
+  [./test_external_pickled]
+    type = 'RavenPython'
+    input = '../../../../scripts/externalROMloader.py load_ROM_externally.xml'
```
Comment (PaulTalbot-INL) [inline tests/framework/ROM/pickleTests/tests]: Did this file make it into the PR?

File: scripts/externalROMloader.py [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+      for el,val in out.items():
+        element      = ET.Element(el)
+        element.text = str(val)
+        realizationData.append(element)
+      evalData.append(realizationData)
+    outElement.append(evalData)
+  with open (outputFilename,"w") as fileObject:
+    fileObject.write(minidom.parseString(ET.tostring(outElement)).toprettyxml(indent="  "))
+
+
```
Comment (PaulTalbot-INL) [inline scripts/externalROMloader.py]: Is there a how-to on using this feature in the guide and/or manual?

File: tests/framework/ROM/pickleTests/tests (hunk 1 above)
Comment (alfoa) [reply tests/framework/ROM/pickleTests/tests]: Yes. it is in the scripts and in the test folder

File: scripts/externalROMloader.py (hunk 2 above)
Comment (alfoa) [reply scripts/externalROMloader.py]: Yes. the documentation has been added in the IOStep section

File: doc/user_manual/step.tex [hunk 3]
```diff
@@ -544,25 +544,76 @@ \subsection{IOStep}
...
-eval=unpickledObj.evaluate(dictionary)
+rom = ravenROMexternal("path_to_pickled_rom/ROM.pk",
+                                     "path_to_RAVEN_framework")
+request = {"x1":np.atleast_1d(Value1),"x2":np.atleast_1d(Value2)}
+eval  = rom.evaluate(request)
 print str(eval)
-fileobj.close()
+\end{lstlisting}
+
+The module above can also be used to evaluate a ROM from input file:
```
Comment (mandd) [inline doc/user_manual/step.tex]: what is the scope of this functionality?

File: scripts/externalROMloader.py [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+#External Modules--------------------end
+
+class ravenROMexternal(object):
+  """
+    External ROM loader and user-gate
+  """
+  def __init__(self, binaryFileName, whereFrameworkIs):
+    """
+      This constructor un-serialize the ROM generated ARMA and
+      make this available for external usage
```
Comment (mandd) [inline scripts/externalROMloader.py]: does it work only for ARMA?

File: scripts/externalROMloader.py [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+from xml.dom import minidom
+#External Modules--------------------end
+
+class ravenROMexternal(object):
+  """
+    External ROM loader and user-gate
+  """
+  def __init__(self, binaryFileName, whereFrameworkIs):
+    """
+      This constructor un-serialize the ROM generated ARMA and
```
[the remark this answers is not in the document]
Comment (alfoa) [reply scripts/externalROMloader.py]: fixed

File: scripts/externalROMloader.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (alfoa) [reply scripts/externalROMloader.py]: fixed

Comment (alfoa) [reply scripts/externalROMloader.py]: No...typo... fixed

File: doc/user_manual/step.tex (hunk 3 above)
Comment (alfoa) [reply doc/user_manual/step.tex]: Just for regression testing basically and to test the functionality in "standalone" mode

**Comment by PaulTalbot-INL** [comment]
Aside from the Windows test failing, this looks good to me.

**Comment by alfoa** [comment]
I will verify

**Comment by alfoa** [comment]
@mandd and @PaulTalbot-INL , it looks now we are green.

**Comment by mandd** [comment]
ready to merge!

**Comment by PaulTalbot-INL** [comment]
I feel like adding the pickle file directly will be a move we reverse someday in the future, but for now I think it's fine.

**Comment by alfoa** [comment]
I added the pickled ROM just to  have a regression test that shows that the serialized object we create are cross-platform compatible (created in an OS but loaded in another)... It should not be a big problem (I hope)

**Comment by PaulTalbot-INL** [comment]
Oh, that's a good check, as well.  Cool.

**Pull request merged**